General sweep of using FString rather than c string

River-Salmon opened the pull request:
This change covers converting the Console, Bindings, and Config apis to using FString rather than manually managing c strings.
Tested with Doom 2 and a pile of mods that stress different parts of the engine (sunlust, Project Malice, Legendoom)

- [x] I have reviewed [CONTRIBUTING.md](../blob/trunk/CONTRIBUTING.md) and agree to its terms.

River-Salmon committed:
Use FString instead of c strings in many use cases:

This change covers converting the Console, Bindings, and Config apis to using FString rather than manually managing c strings.
Tested with Doom 2 and a pile of mods that stress different parts of the engine (sunlust, Project Malice, Legendoom)

River-Salmon committed:
Merge remote-tracking branch 'upstream/main' into strcopy-insecure

RicardoLuis0 commented on src/common/console/c_bind.cpp:
```diff
@@ -190,24 +192,24 @@ static int GetKeyFromName (const char *name)
 	if (keynum == 0)
 	{
-		if (stricmp (key, "LeftBracket") == 0)
+		if (key==FString("LeftBracket"))
```
this should use .CompareNoCase instead of casting, so that stricmp is used directly instead of creating a new string -- plus just using raw `==` is using cased comparison instead of caseless

River-Salmon replied:
Addressed in new commits. Using EqualNoCase now.

RicardoLuis0 commented on src/am_map.cpp:
```diff
@@ -1476,17 +1476,21 @@ bool DAutomap::Responder (event_t *ev, bool last)
-			if (defbind && !strnicmp(defbind, "+am_pan", 7)) return false;
+			static FString amPanString = FString("+am_pan");
+			const FString& defbind = AutomapBindings.GetBind(ev->data1).GetChars();
+			if (!defbind.IsEmpty() && (defbind != amPanString))
```
same for this with .CompareNoCase

RicardoLuis0 commented:
and for anywhere that does stricmp really, this engine looooooooves caseless shit

River-Salmon committed:
Add caseless equals comparison for string use cases

River-Salmon committed:
fixing cases of FString being passed directly to Printf:

Clang does not allow FString to be passed to Printf, must use FString::GetChars() to get a char ptr instead

RicardoLuis0 commented on src/common/console/c_bind.cpp:
```diff
@@ -497,14 +470,14 @@ int FKeyBindings::GetKeysForCommand (const char *cmd, int *first, int *second)
 
 	while (i < NUM_KEYS)
 	{
-		if (stricmp (cmd, Binds[i].GetChars()) == 0)
```
this also needs to be a caseless compare

RicardoLuis0 commented on src/common/console/c_bind.cpp:
```diff
@@ -519,13 +492,13 @@ TArray<int> FKeyBindings::GetKeysForCommand (const char *cmd)
 
 	for (i = 0; i < NUM_KEYS; i++)
 	{
-		if (!stricmp (str, Binds[i].GetChars()))
```
and this

RicardoLuis0 commented on src/common/console/c_bind.cpp:
```diff
@@ -552,7 +525,7 @@ void FKeyBindings::DefaultBind(const char *keyname, const char *cmd)
 	}
 	for (int i = 0; i < NUM_KEYS; ++i)
 	{
-		if (!Binds[i].IsEmpty() && stricmp (Binds[i].GetChars(), cmd) == 0)
```
and this

RicardoLuis0 commented on src/intermission/intermission.cpp:
```diff
@@ -989,21 +989,26 @@ void DIntermissionController::Start()
-			if (cmd != nullptr)
+			if (cmd.Len() > 0)
 			{
-				if (!stricmp(cmd, "toggleconsole") || !stricmp(cmd, "screenshot"))
```
and here

RicardoLuis0 commented on src/intermission/intermission.cpp:
```diff
@@ -989,21 +989,26 @@ void DIntermissionController::Start()
 				}
 				// The following is needed to be able to enter main menu with a controller,
 				// by pressing buttons that are usually assigned to this action, Start and Back by default
-				else if (!stricmp(cmd, "menu_main") || !stricmp(cmd, "pause"))
```
and here

RicardoLuis0 commented on src/gameconfigfile.cpp:
```diff
@@ -981,12 +986,12 @@ void FGameConfigFile::AddAutoexec (FArgs *list, const char *game)
 	{
 		while (NextInSection (key, value))
 		{
-			if (stricmp (key, "Path") == 0 && *value != '\0')
```
and here

RicardoLuis0 commented on src/common/utility/configfile.cpp:
```diff
@@ -531,11 +529,11 @@ void FConfigFile::RenameSection (const char *oldname, const char *newname) const
 {
 	FConfigEntry *probe = section->RootEntry;
 
-	while (probe != NULL && stricmp (probe->Key, key) != 0)
```
and here

RicardoLuis0 commented on src/common/utility/configfile.cpp:
```diff
@@ -382,15 +381,15 @@ bool FConfigFile::DeleteCurrentSection()
 	}
 	FConfigEntry **prober = &CurrentSection->RootEntry, *probe = *prober;
 
-	while (probe != NULL && stricmp(probe->Key, key) != 0)
```
and here

RicardoLuis0 commented:
(and i don't want to be annoying, just trying to get every place where it needs to be caseless 👍)

River-Salmon committed:
more caseless checks for c-string -> Fstring + clang compile fix

-more cases using a caseless string comparison
-fixed another case of FString being passed to Printf directly when clang does not allow that

River-Salmon commented:
Not annoying, useful! I've updated the PR with more fixes.

kcat commented on src/am_map.cpp:
```diff
@@ -1476,17 +1482,21 @@ bool DAutomap::Responder (event_t *ev, bool last)
-			if (defbind && !strnicmp(defbind, "+am_pan", 7)) return false;
+			static FString amPanString = FString();
+			const FString& defbind = AutomapBindings.GetBind(ev->data1).GetChars();
+			if (!defbind.IsEmpty() && (!defbind.IsEqualNoCase(Tokens::am_pan)))
```
I don't think this does the same thing. Other uses seem to indicate `IsEqualNoCase` returns true if they're equal, as opposed to `strnicmp` which returns 0, which means it now takes the branch if they're not equal instead of when they are. Also, the `strnicmp` limited the check to 7 characters, rather than the whole string.

kcat commented on src/am_map.cpp:
```diff
@@ -1476,17 +1482,21 @@ bool DAutomap::Responder (event_t *ev, bool last)
-			const char *defbind = AutomapBindings.GetBind(ev->data1);
-			if (defbind && !strnicmp(defbind, "+am_pan", 7)) return false;
+			static FString amPanString = FString();
+			const FString& defbind = AutomapBindings.GetBind(ev->data1).GetChars();
```
Did you mistakenly add `GetChars()` here? Seems odd to get the chars only to create another `FString` from them, instead of using the `FString` from `GetBind`. Also, `amPanString` is unused.

River-Salmon replied:
updated

River-Salmon committed:
Fixed leftover rot from string refactor in am_map

River-Salmon committed:
Merge remote-tracking branch 'upstream/main' into strcopy-insecure

kcat commented on src/am_map.cpp:
```diff
@@ -1476,17 +1482,20 @@ bool DAutomap::Responder (event_t *ev, bool last)
-			const char *defbind = AutomapBindings.GetBind(ev->data1);
-			if (defbind && !strnicmp(defbind, "+am_pan", 7)) return false;
+			const FString& defbind = AutomapBindings.GetBind(ev->data1);
+			if (!defbind.IsEmpty() && (defbind.IsEqualNoCase(Tokens::am_pan)))
```
Have you checked that this still works the same? This originally only checked the first 7 characters (the length of the `+am_pan` prefix), rather than the whole string. `+am_pandown`, `+am_panup`, etc, is not equal to `+am_pan` if whole strings are considered.

River-Salmon replied:
Good catch. Updated.

River-Salmon committed:
Fix behavior to check for am_pan prefix in automap

rheit commented on src/common/console/c_bind.cpp:
```diff
@@ -165,21 +165,23 @@ static FixedBitArray<NUM_KEYS> DClicked;
 	for (i = 0; i < NUM_KEYS; i++)
 	{
-		if (KeyNames[i] && !stricmp (KeyNames[i], name))
+		if (KeyNames[i].Len() > 0 && (KeyNames[i] == name))
```
This was case insensitive before too.

rheit commented on src/common/console/c_bind.cpp:
```diff
@@ -479,7 +452,7 @@ int FKeyBindings::GetKeysForCommand (const char *cmd, int *first, int *second)
 	while (i < NUM_KEYS && c < 2)
 	{
-		if (stricmp (cmd, Binds[i].GetChars()) == 0)
+		if (cmd == Binds[i])
```
Case insensitive here too.

rheit commented on src/common/console/c_bind.cpp:
```diff
@@ -498,7 +498,7 @@ void FKeyBindings::UnbindACommand (const FString& str)
 	for (i = 0; i < NUM_KEYS; i++)
 	{
-		if (str != Binds[i])
+		if (!str.IsEqualNoCase(Binds[i]))
```
The original was `!stricmp`, which checks if it's equal.

rheit commented on src/common/console/c_buttons.cpp:
```diff
@@ -270,21 +270,20 @@ bool FButtonStatus::ReleaseKey (int keynum)
+	cmd_name = btn_name;
 
-	cmd_name[0] = '+';
+	cmd_name.Insert(0, '+');
```
`FString cmd_name = '+' + btn_name;` would be better because it doesn't need to shift any memory around.

rheit commented on src/common/console/c_buttons.cpp:
```diff
@@ -270,21 +270,20 @@ bool FButtonStatus::ReleaseKey (int keynum)
 
-	cmd_name[0] = '-';
+	cmd_name.Remove(0, 1);
+	cmd_name.Insert(0, '-');
```
You could do `cmd_name.LockBuffer()[0] = '-';` to avoid array shuffling here too.

rheit commented on src/common/console/c_expr.cpp:
```diff
@@ -319,11 +329,11 @@ static FStringProd *NewStringProd (const char *str)
+static FStringProd* NewStringProd (size_t len)
 {
-	FStringProd *prod = (FStringProd *)M_Malloc (sizeof(FStringProd)+len);
+	//FStringProd* prod = (FStringProd *)M_Malloc (sizeof(FStringProd)+len);
```
Did you mean to leave this comment here? For that matter, the memory for the string is no longer allocated as part of the struct, so does this function need to exist anymore?

rheit commented on src/common/console/c_expr.cpp:
```diff
@@ -453,12 +463,12 @@ FProduction *ProdAddDbl (FDoubleProd *prod1, FDoubleProd *prod2)
 	FStringProd *prod = NewStringProd (len);
-	strcpy (prod->Value, prod1->Value);
-	strcat (prod->Value, prod2->Value);
+	prod->Value = NewString;
```
You could simplify and do `prod = NewStringProd(prod1->Value + prod2->Value)` here.

rheit commented on src/common/console/c_expr.cpp:
```diff
@@ -591,7 +601,7 @@ FProduction *ProdNeqDbl (FDoubleProd *prod1, FDoubleProd *prod2)
 FProduction *ProdLTStr (FStringProd *prod1, FStringProd *prod2)
 {
-	return NewDoubleProd (stricmp (prod1->Value, prod2->Value) < 0);
+	return NewDoubleProd (stricmp (prod1->Value.GetChars(), prod2->Value.GetChars()) < 0);
```
Why not use `.CompareNoCase()` here instead of adding `.GetChars()`s? Same with all the following `stricmp`s.

rheit commented on src/common/utility/configfile.cpp:
```diff
@@ -859,7 +855,7 @@ const char *FConfigFile::GenerateEndTag(const char *value)
 		}
 	}
-	while (strstr(value, EndTag) != NULL);
+	while (strstr(value.GetChars(), EndTag) != NULL);
```
You could use `IndexOf()` and avoid `GetChars()`

rheit commented on src/common/utility/configfile.cpp:
```diff
@@ -807,13 +803,13 @@ bool FConfigFile::WriteConfigFile () const
 		while (entry != NULL)
 		{
-			if (strpbrk(entry->Value, "\r\n") == NULL)
+			if (strpbrk(entry->Value.GetChars(), "\r\n") == NULL)
```
You could use `IndexOfAny()` and avoid `GetChars()`

rheit commented on src/gameconfigfile.cpp:
```diff
@@ -981,12 +986,12 @@ void FGameConfigFile::AddAutoexec (FArgs *list, const char *game)
 		while (NextInSection (key, value))
 		{
-			if (stricmp (key, "Path") == 0 && *value != '\0')
+			if (key.IsEqualNoCase(Tokens::path) && !value[0] != '\0')
```
`!value[0] != '\0'` should be `value.IsNotEmpty()`, no?

rheit commented on src/gamedata/a_weapons.cpp:
```diff
@@ -510,15 +511,15 @@ int FWeaponSlots::RestoreSlots(FConfigFile *config, const char *section)
 	while (config->NextInSection (key, value))
 	{
-		if (strnicmp (key, "Slot[", 5) != 0 ||
+		if (key == FString("Slot[") ||
```
This was checking if `key` starts with `"Slot["` (case-insensitive), not if it equals it.

rheit commented:
> and for anywhere that does stricmp really, this engine looooooooves caseless shit

You can thank Hexen for setting that precedent! 😛

River-Salmon committed:
Addressed rheit's fixes for certain string checks

River-Salmon committed:
Addressed more case insensitive tests that I missed

River-Salmon committed:
Merge branch 'trunk' into strcopy-insecure

River-Salmon committed:
fstring sweep - Resolving compile error caused by update to trunk

the-phinet commented:
Head's up: We now require agreeing to our [CONTRIBUTING.md](https://github.com/UZDoom/UZDoom/blob/trunk/CONTRIBUTING.md) file, in order for code to be merged. Since your pr was made before this change, I have manually added the checkbox to your pr's description.
﻿
Basically, the file says: You wrote the code yourself and are releasing it to the project under the gpl
﻿
If you're good with the terms laid out in the file, please click the little checkbox above.
